### PR TITLE
Refine API.spec mechanism

### PR DESCRIPTION
**Pull request by sneaxiy**
### 中文
如何消除API.spec conflict的问题，现在的解决方案是在PR_CI_CPU_Py2 (Compile)的CI任务上做这样的改进：

1. 先编译develop分支的Paddle代码，生成API_DEV.spec
2. 然后checkout到PR的分支，编译PR的代码，生成API_PR.spec
3. 比较API_DEV.spec和API_PR.spec有没有diff，即可知道PR是否造成了API改动（包括代码和文档）。如果检测到有diff，依然需要特定RD approve。

API.spec计划可以从Paddle repo里删除，API_DEV.spec和API_PR.spec都只在CI任务里生成，不存在与Paddle repo，因此这个方案下不会造成两个PR改同一个文件造成conflict。

由于PR_CI_CPU_Py2 (Compile)只编译CPU的包，速度很快。在此方案下（两次编译，develop和PR）CI总时间只增加1min左右（最长增加7min，可能是机器问题），总CI时长14-20min内跑完。

### English
Originally, we use a file named `API.spec` to check whether the PR has revised something to do with API (either API interfaces or document). But the file may be conflicted between 2 PRs which both revise the same file `API.spec`. In fact, this file is not necessary to be inside Paddle repo.

We use a new way to distinguish whether the PR has revised `API.spec` without the real file named `API.spec` existing inside Paddle repo.
- Compile upstream/develop branch to generate `API_DEV.spec`, which represents the current API.spec.
- Compile PR branch to generate `API_PR.spec`, which represents the API.spec generated by the PR revision.
- Find out whether `API_DEV.spec` and `API_PR.spec` have any difference. If yes, the PR needs to be approved by certain people the same as before.

Both `API_DEV.spec` and `API_PR.spec` are generated inside CI tasks, and they would not exist in Paddle repo. Therefore, even though there are 2 PRs which both revise APIs, they would not be conflict with each other.

After API.spec updating, CI Time increases: about 13min -> 14min.

I have changed the document of `layers.relu` and `layers.fc`, and the output of CI is as follows:

CI link: http://ci.paddlepaddle.org/viewLog.html?buildId=196314&buildTypeId=YYTest_PrCiCpuPy2&tab=buildLog

```
[06:10:36]	[Step 1/1] 1 . You must have two RD (XiaoguangHu01 or wanghaoshuang or guoshengCS or heavengate or kuke or Superjomn or lanxianghit or cyj1986 or hutuxian or frankwhzhang or nepeplwu) approval for the api change for the management reason of API interface and API document.
[06:10:36]	[Step 1/1] There are 1 approved errors.
[06:10:36]	[Step 1/1] ****************
[06:10:36]	[Step 1/1] API Difference is: 
[06:10:36]	[Step 1/1] - paddle.fluid.layers.fc (ArgSpec(args=['input', 'size', 'num_flatten_dims', 'param_attr', 'bias_attr', 'act', 'name'], varargs=None, keywords=None, defaults=(1, None, None, None, None)), ('document', 'e28421f1253a3545d9bfe81a8028ea68'))
[06:10:36]	[Step 1/1] ?                                                                                                                                                                                                         ^  ------------------ ^^^^^^^^^^^^^
[06:10:36]	[Step 1/1] 
[06:10:36]	[Step 1/1] + paddle.fluid.layers.fc (ArgSpec(args=['input', 'size', 'num_flatten_dims', 'param_attr', 'bias_attr', 'act', 'name'], varargs=None, keywords=None, defaults=(1, None, None, None, None)), ('document', '3385a2dbe27701f33075f7141fc088c7'))
[06:10:36]	[Step 1/1] ?                                                                                                                                                                                                         ^^^^^ ++++++++++++++++++++++  ^^^^^
[06:10:36]	[Step 1/1] 
[06:10:36]	[Step 1/1] - paddle.fluid.layers.relu (ArgSpec(args=['x', 'name'], varargs=None, keywords=None, defaults=(None,)), ('document', 'c46011e3d848bc3dc650772b25fbbf10'))
[06:10:36]	[Step 1/1] ?                                                                                                                     ^^^   ^     - ^^^^^^^^^^^^  ^^^
[06:10:36]	[Step 1/1] 
[06:10:36]	[Step 1/1] + paddle.fluid.layers.relu (ArgSpec(args=['x', 'name'], varargs=None, keywords=None, defaults=(None,)), ('document', '55f015781ccb053d86f2d94f8ca8fb00'))
[06:10:36]	[Step 1/1] ?                                                                                                                     ^^^  +++ ^^^^^   +++++ +  ^^  ^
[06:10:36]	
```

![a441c05d88f342bad49a6823a0d8e7be](https://user-images.githubusercontent.com/32832641/66734192-b9b92580-ee94-11e9-828d-1a733bb5f2a4.jpg)

### CODE REVIEW

File: tools/diff_api.py [hunk 1]
```diff
@@ -15,21 +15,17 @@
...
-    print(
-        '''If you modify/add/delete the API files, including code and comment, please follow these steps in order to pass the CI:
-    1. cd ${paddle_path}, compile paddle;
-    2. pip install build/python/dist/(build whl package);
-    3. run "python tools/print_signatures.py paddle.fluid> paddle/fluid/API.spec"'''
-    )
-    sys.exit(1)
+    print('API Difference is: ')
+    for each_diff in diffs:
+        print(each_diff)
```
Comment (luotao1) [inline tools/diff_api.py:38]: 1，2，3的步骤可以写在注释里，不然以后就不知道api.spec怎么生成了。

File: .gitignore [hunk 2]
```diff
@@ -5,6 +5,8 @@ python/paddle/v2/fluid/tests/book/image_classification_resnet.inference.model/
 python/paddle/v2/fluid/tests/book/image_classification_vgg.inference.model/
 python/paddle/v2/fluid/tests/book/label_semantic_roles.inference.model/
 paddle/fluid/operators/distributed/send_recv.proto
+paddle/fluid/API_DEV.spec
+paddle/fluid/API_PR.spec
```
Comment (luotao1) [inline .gitignore:4]: 方便的话，可以把ignore里的1-6行去掉。

File: tools/check_api_approvals.sh [hunk 3]
```diff
@@ -42,6 +42,14 @@ if [[ $git_files -gt 19 || $git_count -gt 999 ]];then
...
 fi    
 
+api_spec_diff=`python ${PADDLE_ROOT}/tools/diff_api.py ${PADDLE_ROOT}/paddle/fluid/API_DEV.spec  ${PADDLE_ROOT}/paddle/fluid/API_PR.spec` 
+if [ "$api_spec_diff" != "" ]; then
+    APPROVALS=`echo ${approval_line}|python ${PADDLE_ROOT}/tools/check_pr_approval.py 2 46782768 7534971 14105589 12605721 3064195 328693 47554610 39645414 11195205 20274488 45024560     `
+    failed_num=`expr $failed_num + 1`
+    echo_line="You must have two RD (XiaoguangHu01 or wanghaoshuang or guoshengCS or heavengate or kuke or Superjomn or lanxianghit or cyj1986 or hutuxian or frankwhzhang or nepeplwu) approval for the api change for the management reason of API interface and API document."
+    echo_list=(${echo_list[@]}$failed_num "." $echo_line)
+fi
+
```
Comment (luotao1) [inline tools/check_api_approvals.sh:51]: 46-49行和后面重复了。

File: .gitignore (hunk 2 above)
Comment (sneaxiy) [reply .gitignore:4]: Done.

File: tools/check_api_approvals.sh (hunk 3 above)
Comment (sneaxiy) [reply tools/check_api_approvals.sh:51]: Done.

File: tools/diff_api.py (hunk 1 above)
Comment (sneaxiy) [reply tools/diff_api.py:38]: Done.

**Review by luotao1 — DISMISSED** [review]
LGTM

**Review by luotao1 — DISMISSED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM

**Pull request merged**